File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* findKthNode(ListNode* temp, int k){
        if (k == 0) return temp;
        while (temp != nullptr && k > 1) {  
            temp = temp->next;
            k--;
        }
        return temp;
    }

    ListNode* reverse(ListNode* head){
        ListNode* temp = head;
        ListNode* prev = nullptr;
        ListNode* nextEl = nullptr;

        while (temp != nullptr){
            nextEl = temp->next;
            temp->next = prev;
            prev = temp;
            temp = nextEl;
        }

        return prev;
    }

    ListNode* reverseKGroup(ListNode* head, int k) {
        if (head == nullptr) return nullptr;

        ListNode* temp = head;
        ListNode* newHead = nullptr;       
        ListNode* prevTail = nullptr;      

        while (temp != nullptr) {
            ListNode* KthNode = findKthNode(temp, k);

            if (KthNode == nullptr) {
                if (prevTail != nullptr) {
                    prevTail->next = temp; 
                }
                break;
            }

            ListNode* nextNode = KthNode->next;  

            KthNode->next = nullptr;             
            ListNode* revHead = reverse(temp);   

            if (newHead == nullptr) {
                newHead = revHead;               
            }

            if (prevTail != nullptr) {
                prevTail->next = revHead;       
            }

            prevTail = temp;                     
            temp = nextNode;                     
        }

        return newHead;
    }
};

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.py
class Solution:
    def findKthNode(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        while head != None and k > 1:
            head = head.next
            k = k - 1
        return head
    
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        curr = head
        prev = None

        while curr != None:
            nextEl = curr.next
            curr.next = prev
            prev = curr
            curr = nextEl
        
        return prev

    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        temp = head
        prevEle = None
        newHead = None

        while temp != None:
            kthElement = self.findKthNode(temp, k)
            
            if kthElement == None:
                if prevEle:
                    prevEle.next = temp
                break
            
            nextEle = kthElement.next
            kthElement.next = None
            
            reverseHead = self.reverseList(temp)
            
            if temp == head:
                newHead = reverseHead
            
            if prevEle:
                prevEle.next = reverseHead
            
            prevEle = temp
            temp = nextEle
        
        return newHead if newHead else head
